src/Servers/Kestrel/Core/src/Internal/HttpsConnectionAdapter.cs

```diff
@@ -195,6 +197,13 @@ private async Task<IAdaptedConnection> InnerOnConnectionAsync(ConnectionAdapterC
             }
             finally
             {
+                if (!sslStream.IsAuthenticated && (_serverCertificate != null ||
+                    CertificateManager.IsHttpsDevelopmentCertificate(_serverCertificate) ||
+                    !CertificateManager.CheckDeveloperCertificateKey(_serverCertificate)))
```

Comment:
Why can't this be checked at startup when we first retrieve the dev cert? Why do you have to wait for a connection to fail? People are more likely to see the error it it's at the top of the logs.

Reply:
@Tratcher that's how we do it in 3.1.

Reply:
This bit I'm not super sure about, as the code is vastly different from 3.0/3.1. I would appreciate if @halter73 or you help me make sure this is the right check/fix at the right place.

Reply:
See https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs#L221-L236 for details.

Reply:
Ah, didn't notice this was a 2.1 backport. The question stands though; is CheckDeveloperCertificateKey reliable enough that we can check it without first waiting for the SslStream auth to fail?

Reply:
Also, to mimic the 3.1 code this should at least be moved up into the prior catch block where AuthenticationException is handled.

Reply:
It seems to behave differently in 2.1 than in 3.1, that's why I'm not 100% sure about the change. There are some tests failing on the CI, so I'll have to refine it tomorrow.

Reply:
This boolean logic seems backwards. It's only calling into the CertificateManager manager `_serverCertificate == null` because the `||` short-circuits when `_serverCertificate != null` is true.

Why not move this logic into the `catch (Exception ex) when (ex is IOException || ex is AuthenticationException)` block and use that same if condition as in master but with `ex is AuthenticationException &&` prefixed?

Reply:
I figured this out. In 2.1 is throwing a different exception type 😵